### ISSUE TITLE
ingest pre-processed tasmax/tasmin (not from Pangeo CMIP6 archive) 

### ISSUE THREAD

**Issue by dgergel** [ClimateImpactLab/downscaleCMIP6#426]
For very limited models (just the GFDL* models as of now), we are pre-processing them to avoid having negative DTR. We need to have the ability to ingest these models from a GCS location other than the Pangeo CMIP6 archive and add that to the e2e workflows.

**Comment by brews** [comment]
If you have a fsspec-style URL, the fastest/dirtiest way to do this is just to transfer it over into the "raw" storage as though it were GFDL* from the CMIP6 catalog (use the same path as we had for other downloads). It needs the usual formatting, etc, as though it were raw -- naturally. This "raw" storage is where cleaning workflows will pick it up and begin work. By default, the initial "download CMIP6" step won't overwrite the existing data (assuming the paths match).

We already have a [workflowtemplate to transfer zarr stores between URLs](https://github.com/ClimateImpactLab/downscaleCMIP6/blob/master/workflows/templates/zarr-transfer.yaml). So we could write a dumb ol' workflow (not a workflowtemplate). We could call it "workflows/transfer_special_gfdl.yaml" with a comment at the top of it explaining what it does. This workflow would just use `workflows/templates/zarr-transfer.yaml` with your input URLs and the output URLs to raw already fixed.

How does that sound, @dgergel ?

Actually, this might be a good one for @emileten if he has the time and wants to do more raw kubernetes/argo/cloud engineering stuff.

**Comment by emileten** [comment]
@brews I'm in ! A clarification question though. I understood @dgergel wants to do this transformation in a notebook. Can't she save the output directly to that 'raw' storage you mention from within that notebook ? If so, why would we use argo ?

**Comment by brews** [comment]
@emileten, that's a good point. 

@dgergel ? I think you already have the permissions for this.

**Comment by emileten** [comment]
We would use argo if we were to do this transformation through `dodola` I guess? But the notebook is faster?

**Comment by dgergel** [comment]
@brews I think I already have permissions for that. Will check today. 

I was thinking it was easier to do in a notebook because it's just a simple tasmax/tasmin switch, and tasmax runs don't require tasmin specs, so I figured this way we don't have to make any workflow changes.

**Comment by brews** [comment]
@dgergel Cool. Let me know if you need any permission changes. 

I'm fine with a notebook solution so we can get this done. 

Let me know if you need help working out the paths you'd need to write to.

As an example, [the last GFDL-ESM4 tasmax run we did](https://argo.cildc6.org/archived-workflows/default/e3eceb83-7724-4740-927f-74cd02813374) wrote to:

- gs://raw-305d04da/cmip6/CMIP/NOAA-GFDL/GFDL-ESM4/historical/r1i1p1f1/day/tasmax/gr1/v20190726.zarr
- gs://raw-305d04da/cmip6/ScenarioMIP/NOAA-GFDL/GFDL-ESM4/ssp370/r1i1p1f1/day/tasmax/gr1/v20180701.zarr
- gs://raw-305d04da/cmip6/ScenarioMIP/NOAA-GFDL/GFDL-ESM4/ssp245/r1i1p1f1/day/tasmax/gr1/v20180701.zarr
- gs://raw-305d04da/cmip6/ScenarioMIP/NOAA-GFDL/GFDL-ESM4/ssp126/r1i1p1f1/day/tasmax/gr1/v20180701.zarr

And this is where the cleaning step picks up the raw data and continues on...

I figure once you've transferred the files, we could then run the workflow and ensure it grabs and runs as we expect it to.

How does that sound as an action plan, @dgergel ?

**Comment by dgergel** [comment]
Sounds good @brews. 

So to clarify, you're suggesting I write to the _same_ GFDL-ESM4 paths as those above, correct? And overwrite the existing files in raw? 

Then for GFDL-CM4, is there a workflow id that I should use? I browsed the archived workflows and didn't see one for GFDL-CM4 but maybe I missed it.

**Comment by brews** [comment]
@dgergel The paths above are from a past run, I posted them to demonstrate how/where the raw zarr files are written. I don't know what GFDL* data you will be modifying from (I mean, we've modified the parameter files since we opened the issue) so it will change based on that.

Written as a python f-string the pattern for the "raw data" URLs is:

```python
f"gs://raw-305d04da/cmip6/{activity_id}/{institution_id}/{source_id}/{experiment_id}/{member_id}/{table_id}/{variable_id}/{grid_label}/v{version}.zarr"
```

This URL pattern is what the "download" step always writes to. The "cleaning" step picks up data from the same URL pattern. Argo gets this information from the parameter file you're using for the run (in this case, the GFDL* ones).

The "download" step does nothing but read the data and then write it. The data is "raw". It has no workflow ID in its metadata. No metadata is adjusted. The "version" used is the version given by cmip6 in the cloud, not our dc6_version. Your notebook is going to be the one exception to that because its' swapping tasmin and tasmax but other than that it should look exactly like the raw data we would get otherwise.

Your notebook is essentially taking the role of the "download" step. So you'd grab the information from the parameter files to read from the CMIP6-in-the-cloud. Do whatever tasmin/tasmax swapping is needed, and then write the corrected data to `gs://raw-305d04da/` using the above URL pattern.  If you're sure you've got the paths right, I think it's okay to overwrite what's there.

So, in short, look to the GFDL parameter files in workflows/parameters and that information with the above URL pattern to get the URL you'd write to.

How's that?

**Comment by dgergel** [comment]
Yep - what you're saying here is what I'd already outlined in previous slack conversations on this topic but thanks for reiterating ;-) 

Was just making sure that the `raw-305d04da` was not unique to a model run, and now I get that it isn't

**Comment by brews** [comment]
@dgergel Gotcha. Nope. That's not unique to the GCM.

Bucket names need to be globally unique across all projects on GCS so that's why our bucket names look a little funny.

**Comment by brews** [comment]
No further work is needed on the part of workflows so I'm going to close this.
